bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceOverlay.java

```diff
@@ -0,0 +1,1044 @@
+		regexSearchButton.setSelection(findReplaceLogic.isActive(SearchOptions.REGEX));
+		regexSearchButton.addSelectionListener(new SelectionListener() {
+
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				activateInFindReplacerIf(SearchOptions.REGEX, ((ToolItem) e.widget).getSelection());
+				wholeWordSearchButton.setEnabled(!findReplaceLogic.isActive(SearchOptions.REGEX));
+				updateIncrementalSearch();
+			}
+
+			@Override
+			public void widgetDefaultSelected(SelectionEvent e) {
+				// Do nothing
+			}
+		});
+	}
+
+	private void createCaseSensitiveButton() {
+		caseSensitiveSearchButton = new ToolItem(searchTools, SWT.CHECK);
+		caseSensitiveSearchButton.setImage(FindReplaceOverlayImages.get(FindReplaceOverlayImages.OBJ_CASE_SENSITIVE));
+		caseSensitiveSearchButton.setToolTipText(FindReplaceMessages.FindReplaceOverlay_caseSensitiveButton_toolTip);
+		caseSensitiveSearchButton.setSelection(findReplaceLogic.isActive(SearchOptions.CASE_SENSITIVE));
+		caseSensitiveSearchButton.addSelectionListener(new SelectionListener() {
+
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				activateInFindReplacerIf(SearchOptions.CASE_SENSITIVE, caseSensitiveSearchButton.getSelection());
+				updateIncrementalSearch();
+			}
+
+			@Override
+			public void widgetDefaultSelected(SelectionEvent e) {
+				// Do Nothing
+			}
+		});
+	}
+
+	private void createWholeWordsButton() {
+		wholeWordSearchButton = new ToolItem(searchTools, SWT.CHECK);
+		wholeWordSearchButton.setImage(FindReplaceOverlayImages.get(FindReplaceOverlayImages.OBJ_WHOLE_WORD));
+		wholeWordSearchButton.setToolTipText(FindReplaceMessages.FindReplaceOverlay_wholeWordsButton_toolTip);
+		wholeWordSearchButton.setSelection(findReplaceLogic.isActive(SearchOptions.WHOLE_WORD));
+		wholeWordSearchButton.addSelectionListener(new SelectionListener() {
+
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				activateInFindReplacerIf(SearchOptions.WHOLE_WORD, wholeWordSearchButton.getSelection());
+				updateIncrementalSearch();
+			}
+
+			@Override
+			public void widgetDefaultSelected(SelectionEvent e) {
+				// Do Nothing
+			}
+		});
+	}
+
+	private void createReplaceTools() {
+		Color warningColor = JFaceColors.getErrorText(getShell().getDisplay());
+
+		replaceTools = new ToolBar(replaceContainer, SWT.HORIZONTAL);
+		GridDataFactory.fillDefaults().grab(false, true).align(GridData.CENTER, GridData.END).applyTo(replaceTools);
+		replaceButton = new ToolItem(replaceTools, SWT.PUSH);
+		replaceButton.setImage(FindReplaceOverlayImages.get(FindReplaceOverlayImages.OBJ_REPLACE));
+		replaceButton.setToolTipText(FindReplaceMessages.FindReplaceOverlay_replaceButton_toolTip);
+		replaceButton.addSelectionListener(new SelectionListener() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				if (getFindString().isEmpty()) {
+					showUserFeedback(warningColor, true);
+					return;
+				}
+				performSingleReplace();
```

Comment:
`performSingleReplace` is implemented like this:

```java
	private void performSingleReplace() {
		findReplaceLogic.performReplaceAndFind(getFindString(), getReplaceString());
		evaluateFindReplaceStatus();
	}
```
To answer your question: yes, it is necessary - but already done correctly!

This might hint at bad naming, I will rename the method to `performSingleReplaceAndEvaluate`.
Ultimately, we have a wider problem to fix down the line; Somehow the actions for "doing find/replace" and "evaluating status" are tied very closely together. However, this isn't reflected in the code, where we always need to perform two operations (which really should be a single operation!)

Ideally, we would have some kind of observer-pattern which just listens to state-changes in the logic

Reply:
https://github.com/eclipse-platform/eclipse.platform.ui/commit/67ef99dd0640b16d57511fe574efb5974ee221e9

Reply:
Then this is inconsistent throughout the class, as the other `perform...()` method don't do that evaluation but require an additional evaluation step. Why is `performSingleReplace()` different? In addition, the `performReplaceOnEnter()` method now calls the `performSingleReplaceAndEvaluate()` method and still calls the `evaluateFindReplaceStatus()` method afterwards. Even if this requires some general improvement, the first contributed version should at least be consistent itself.

Reply:
I have removed all status-checks inside of these helper methods and delegated them to the callers. I agree with you that having these checks inside of the functions is inconsistent because some methods just don't use such helper functions. That leads to a system where some button-listeners need to update the status and some don't (because it is done in a "perform ..." method).